web/tests/scheduled_messages.test.js

```diff
@@ -0,0 +1,114 @@
+"use strict";
+
+const {strict: assert} = require("assert");
+
+const {zrequire} = require("./lib/namespace");
+const {run_test} = require("./lib/test");
+
+const scheduled_messages = zrequire("scheduled_messages");
+
+function get_expected_send_opts(expecteds) {
+    const modal_opts = {
+        send_later_tomorrow: {
+            tomorrow_nine_am: {
+                text: "translated: Tomorrow at 9:00 AM",
+                time: "9:00 am",
+            },
+            tomorrow_four_pm: {
+                text: "translated: Tomorrow at 4:00 PM",
+                time: "4:00 pm",
+            },
+        },
+        send_later_custom: {
+            text: "translated: Custom",
+        },
+        possible_send_later_today: false,
+        possible_send_later_monday: false,
+    };
+    const optional_modal_opts = {
+        send_later_today: {
+            today_nine_am: {
+                text: "translated: Today at 9:00 AM",
+                time: "9:00 am",
+            },
+            today_four_pm: {
+                text: "translated: Today at 4:00 PM",
+                time: "4:00 pm",
+            },
+        },
+        send_later_monday: {
+            monday_nine_am: {
+                text: "translated: Monday at 9:00 AM",
+                time: "9:00 am",
+            },
+        },
+    };
+
+    // 'today_nine_am'
+    // 'today_four_pm'
+    // 'monday_nine_am'
+    for (const expect of expecteds) {
+        const day = expect.split("_")[0]; // "today", "monday"
+        if (!modal_opts[`possible_send_later_${day}`]) {
+            modal_opts[`possible_send_later_${day}`] = {};
+        }
+        modal_opts[`possible_send_later_${day}`][expect] =
+            optional_modal_opts[`send_later_${day}`][expect];
+    }
+
+    return modal_opts;
+}
+
+run_test("scheduled_modal_opts", () => {
+    // Sunday thru Saturday
+    const days = [
+        "2023-04-30",
+        "2023-05-01",
+        "2023-05-02",
+        "2023-05-03",
+        "2023-05-04",
+        "2023-05-05",
+        "2023-05-06",
+    ];
+    // Extra options change based on the hour of day
+    const options_by_hour = [
+        {hour: "T06:00:00", extras: ["today_nine_am", "today_four_pm"]},
+        {hour: "T11:00:00", extras: ["today_four_pm"]},
+        {hour: "T17:00:00", extras: []},
+    ];
+
+    // Now we can test those hourly options on each day of the week
+    for (const day of days) {
+        for (const opts of options_by_hour) {
+            const date = new Date(day + opts.hour);
+            // On Fridays (5) and Saturdays (6), add the Monday option
+            if (date.getDay() > 4) {
+                opts.extras.push("monday_nine_am");
+            }
+            const modal_opts = scheduled_messages.get_filtered_send_opts(date);
+            const expected_opts = get_expected_send_opts(opts.extras);
+            assert.deepEqual(modal_opts, expected_opts);
+        }
+    }
+});
```

Comment:
I'm not sure it's worth redoing this again (I'll leave that up to you), but I'll comment a bit on testing philosophy here, since the general patterns are important even if the perfect tests for this specific feature are not. Generally, the best tests are those that don't have their own logic, even if it means a bit of copy-pasting. The main reason is that the more code/logic you have in your tests, the more likely it is that you have a bug in that code. But equally important is that tests written the right way can serve as excellent self-documentation of the expected behavior in the cases of interest, and if we make a change to the logic later, it'll be incredibly obvious from the PR that it changed the output of, say, just the Thursday test, removing the temptation to do manual testing to verify the logic (one would still review the code and do manual testing to verify the visuals).

For example, in this case, I think the best possible test suite for something like this would look like this:
```
// Sunday
assert_expected_options("2023-04-30 06:00:00", ["today_nine_am", "today_four_pm"] + tomorrow_options);
assert_expected_options("2023-04-30 11:00:00", ["today_four_pm"] + tomorrow_options);
assert_expected_options("2023-04-30 17:00:00", tomorrow_options);
// Monday
...
```

This high-level topic might be a topic worth adding something to 
https://zulip.readthedocs.io/en/latest/testing/philosophy.html about.

Reply:
I really appreciate your taking the time to point all of this out, Tim. I'm happy to both rewrite these tests in line with your suggestions here, and to update the testing-philosophy docs. I can even imagine drafting an update to them with specific references/examples for this set of fixes.

If you'd like to see the fixes themselves as part of this PR, I'm happy to do that. Otherwise, I'll just add this to my list and open a subsequent PR in a day or two.

Reply:
Yeah, I think those can go on the TODO list, and we can merge this and check off the release goal.